swri_image_util/src/nodelets/replace_colors_nodelet.cpp

```diff
@@ -0,0 +1,457 @@
+    // Publishes the modified image
+    image_transport::Publisher image_pub_;
+    // Subscribes to the original image
+    image_transport::Subscriber image_sub_;
+    // Mapping from a colormap name to the OpenCV integer representation
+    std::map<std::string, int32_t> colormap_names_;
+  };
+
+  ReplaceColorsNodelet::ReplaceColorsNodelet()
+  {
+    // Initialize the colormap name mapping. Every OpenCV colormap should have
+    // a string identifiying it. This allows the node to easily take a user
+    // parameter and convert it to a representation the algorithm can use.
+    // OpenCV 2.x does not have the Parula colormap, so only include this
+    // when OpenCV 3.x is available
+    colormap_names_["autumn"] = cv::COLORMAP_AUTUMN;
+    colormap_names_["bone"] = cv::COLORMAP_BONE;
+    colormap_names_["jet"] = cv::COLORMAP_JET;
+    colormap_names_["winter"] = cv::COLORMAP_WINTER;
+    colormap_names_["rainbow"] = cv::COLORMAP_RAINBOW;
+    colormap_names_["ocean"] = cv::COLORMAP_OCEAN;
+    colormap_names_["summer"] = cv::COLORMAP_SUMMER;
+    colormap_names_["spring"] = cv::COLORMAP_SPRING;
+    colormap_names_["cool"] = cv::COLORMAP_COOL;
+    colormap_names_["hsv"] = cv::COLORMAP_HSV;
+    colormap_names_["pink"] = cv::COLORMAP_PINK;
+    colormap_names_["hot"] = cv::COLORMAP_HOT;
+    #if CV_MAJOR_VERSION == 3
+    colormap_names_["parula"] = cv::COLORMAP_PARULA;
+    #endif
+  }
+
+  ReplaceColorsNodelet::~ReplaceColorsNodelet()
+  {
+    // Nothing to clean up on destruction
+  }
+
+  // Initialize the nodelet
+  void ReplaceColorsNodelet::onInit()
+  {
+    // Node handles for interacting with ROS
+    ros::NodeHandle& nh = getNodeHandle();
+    ros::NodeHandle& priv_nh  = getPrivateNodeHandle();
+
+    // Lookup table to replace colors with. By default will just convert the
+    // gray scale values to their RGB equivalents. If this node is ever extended
+    // to more than gray scale, this will have to be changed
+    color_lut_ = cv::Mat::zeros(1, NUM_GRAY_VALUES, CV_8UC3);
+    initLut();
+
+    // Get the array representing the color mapping from the parameter server.
+    // This will be in the format [[u0, [r0, g0, b0]], [u1, [r1, g1, b1]]]
+    // where u is the gray scale value to replace, and r, g, b are the RGB
+    // components to replace it with.
+    XmlRpc::XmlRpcValue color_param;
+
+    // This node has two different methods of changing gray scale values to
+    // color imagery. The first maps the gray scale values to OpenCV colormaps
+    // This call checks for that option
+    bool colormap_specified = priv_nh.getParam("colormap", color_param);
+    if (colormap_specified)
+    {
+      //  Use the colormap the user has requested
+      readColormap(color_param);
+    }
+
+    // The other option for modifying the grayscale images is to define a
+    // lookup table that maps grayscale levels to user defined RGB values.
+    // This can be used in conjunction with the colormap option to replace
```

Comment:
This particular sentence is excellent. I was just thinking, "what happens when the user specifies both?" and here I get my answer. 👍